### ISSUE TITLE
Building from source, gcc issues

### ISSUE THREAD

**Issue by vmayoral** [tensorflow/tensorflow#349]
Hi everyone,

I'm trying to build tensorflow from source but got stucked at the following:

```
root@ubuntu:/tensorflow# ../bazel/output/bazel build -c opt //tensorflow/tools/pip_package:build_pip_package
WARNING: Sandboxed execution is not supported on your system and thus hermeticity of actions cannot be guaranteed. See http://bazel.io/docs/bazel-user-manual.html#sandboxing for more information. You can turn off this warning via --ignore_unsupported_sandboxing.
INFO: Found 1 target...
INFO: From Compiling tensorflow/core/kernels/cwise_op_add.cc:
gcc: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-4.8/README.Bugs> for instructions.
ERROR: /tensorflow/tensorflow/core/BUILD:210:1: C++ compilation of rule '//tensorflow/core:kernels' failed: gcc failed: error executing command /usr/bin/gcc -U_FORTIFY_SOURCE '-D_FORTIFY_SOURCE=1' -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 -DNDEBUG -ffunction-sections ... (remaining 72 argument(s) skipped): com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 4.
Target //tensorflow/tools/pip_package:build_pip_package failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 126.561s, Critical Path: 123.83s
```

`gcc` should be fine:
```
root@ubuntu:/tensorflow# gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/4.8/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 4.8.4-2ubuntu1~14.04' --with-bugurl=file:///usr/share/doc/gcc-4.8/README.Bugs --enable-languages=c,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-4.8 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --with-gxx-include-dir=/usr/include/c++/4.8 --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --enable-gnu-unique-object --disable-libmudflap --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-4.8-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-4.8-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-4.8-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 4.8.4 (Ubuntu 4.8.4-2ubuntu1~14.04) 
```

Every time i execute `../bazel/output/bazel build -c opt //tensorflow/tools/pip_package:build_pip_package` i get the same:
```
WARNING: Sandboxed execution is not supported on your system and thus hermeticity of actions cannot be guaranteed. See http://bazel.io/docs/bazel-user-manual.html#sandboxing for more information. You can turn off this warning via --ignore_unsupported_sandboxing.
INFO: Found 1 target...
[810 / 839] Still waiting for 27 jobs to complete:
      Running (spawn):
        Compiling tensorflow/core/kernels/transpose_op.cc, 10 s
        Compiling tensorflow/core/kernels/cwise_op_add.cc, 10 s
        Compiling tensorflow/core/kernels/slice_op.cc, 10 s
        Compiling tensorflow/core/kernels/cwise_op_equal_to.cc, 10 s
      Scheduling:
        Compiling tensorflow/core/kernels/argmax_op.cc, 10 s
        Compiling tensorflow/core/kernels/cwise_op_sub.cc, 10 s
        Compiling tensorflow/core/kernels/reduction_ops_mean.cc, 10 s
        Compiling tensorflow/core/kernels/reduction_ops_prod.cc, 10 s
        Compiling tensorflow/core/kernels/matmul_op.cc, 10 s
        Compiling tensorflow/core/kernels/reverse_sequence_op.cc, 10 s
        Compiling tensorflow/core/kernels/segment_reduction_ops.cc, 10 s
        Compiling tensorflow/core/kernels/aggregate_ops.cc, 10 s
        Compiling tensorflow/core/kernels/cwise_op_mul.cc, 10 s
        ... 14 more jobs

```

Those 27 missing packages never seem to finish. Any advice here?

**Comment by malfet** [comment]
You are likely running out of memory. Try reducing number of parallel builds by passing '--local_resources  2048,.5,1.0', which would instruct bazel to spawn no more than one compiler process at the time.

**Comment by vmayoral** [comment]
That did it. Thanks @malfet!

**Comment by tobegit3hub** [comment]
Thanks @malfet and @vmayoral 👍

**Comment by dat-ai** [comment]
An other solution is to limit the maximum jobs bazel should build. For example:

```shell
bazel build -c opt --jobs 1 --local_resources 2048,0.5,1.0 --verbose_failures --config=cuda //tensorflow/tools/pip_package:build_pip_package
```

**Comment by yejw5** [comment]
Thanks @malfet @dat-ai

**Comment by loisaidasam** [comment]
> You are likely running out of memory. Try reducing number of parallel builds by passing '--local_resources 2048,.5,1.0', which would instruct bazel to spawn no more than one compiler process at the time.

Worth noting that`--local-resources` refers to:

> comma-separated available amount of RAM (in MB), CPU (in cores) and available I/O (1.0 being average workstation

via https://docs.bazel.build/versions/master/command-line-reference.html#flag--local_resources

**Comment by AmmarkoV** [comment]
> You are likely running out of memory. Try reducing number of parallel builds by passing '--local_resources 2048,.5,1.0', which would instruct bazel to spawn no more than one compiler process at the time.

Thank you very much, bazel is a terrible build system https://archive.fosdem.org/2018/schedule/event/how_to_make_package_managers_cry/
